streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -50,65 +57,135 @@
 import java.util.Map;
 
 public class RecordCollectorImpl implements RecordCollector {
```

Comment:
I am not convinced atm, that the record collector should handle on the committing. IMHO, the `StreamThread` should unify this logic (especially with KIP-447 in mind)-- might be best to discuss in person.

Reply:
Yeah we can discuss this in person definitely. When I was working on this I do not have the full picture of KIP-447 in mind yet. After reviewed your PR I think one way was to let the Record Collector be a per-thread object than per-task, and within the class it would maintain one or multiple producers, the benefits of this is to still maintain the committing within a task --- but again, this is just one way of doing it and I'm open for other options especially that in KIP-447 we think it is better to *always* commit all tasks together.

If we want to push the management of embedded clients to stream-thread while always committing all tasks when needed, maybe we can consider this:

* Let the task.commit() function to only handle the procedure of flushing / checkpointing state stores.
* And then at StreamThread we call `foreach (task) -> commit` and then make another `commit` call which did similar logic as in RecordCollector's commit logic, branching on EOS-alpha / EOS-beta / non-EOS. And also let the StreamThread to keep a reference of all the producer(s) in order to make that call.

By doing this, only per-task RecordCollector and StreamThread would have reference to the producer (and StreamThread have reference to the consumer), and the former only used it to call `produce / flush`, while the latter used it for `commit / close` (the lifecycle of the producers are also managed by the thread).

Let's chat more in person about this.

Reply:
That align with my thinking -- in fact, we could rename `task.commit()` to something like `prepareCommit()`? A task must for sure flush it's state -- not sure if it's need to flush the record collector -- flushing the record collector could be done by the stream-thread. This would decouple tasks even more (and IMHO better) from topics.

Reply:
`prepareCommit()` sounds better.

Regarding the record collector: we need to flush the collector before we flush the state manager due to the following order:

1) flush the state mgr which may generate records to send
2) record-collector.flush -> producer.flush which would update the offset
3) write the offsets to the checkpoint file.
4) then commit.

1/2/3 would be in `prepareCommit` in that order, and then 4) would be in stream-thread. With EOS-enabled, now if 4) failed I would wipe out the state-dir in closeDirty so that the checkpoint files would be wiped as well.